packages/protocol/contracts/common/Accounts.sol

```diff
@@ -284,40 +352,196 @@ contract Accounts is
    * @dev v, r, s constitute `signer`'s signature on `msg.sender`.
    */
   function authorizeAttestationSigner(address signer, uint8 v, bytes32 r, bytes32 s) public {
-    Account storage account = accounts[msg.sender];
-    authorize(signer, v, r, s);
-    account.signers.attestation = signer;
+    authorizeSignerWithSignature(signer, AttestationSigner, v, r, s);
+    setDefaultSigner(signer, AttestationSigner);
+
     emit AttestationSignerAuthorized(msg.sender, signer);
   }
 
+  /**
+   * @notice Begin the process of authorizing an address to sign on behalf of the account
+   * @param signer The address of the signing key to authorize.
+   * @param role The role to authorize signing for.
+   */
+  function authorizeSigner(address signer, string memory role) public {
+    require(isAccount(msg.sender), "Unknown account");
+    require(
+      isNotAccount(signer) && isNotAuthorizedSignerForAnotherAccount(signer),
+      "Cannot re-authorize address signer"
+    );
+
+    accounts[msg.sender].signerAuthorizations[role][signer] = SignerAuthorization({
+      started: true,
+      completed: false
+    });
+
+    emit SignerAuthorized(msg.sender, signer, role);
+  }
+
+  /**
+   * @notice Finish the process of authorizing an address to sign on behalf of the account. 
+   * @param _account The address of account that authorized signing.
+   * @param role The role to finish authorizing for.
+   */
+  function completeSignerAuthorization(address _account, string memory role) public {
+    Account storage account = accounts[_account];
+    require(
+      account.signerAuthorizations[role][msg.sender].started == true,
+      "Signer authorization not started"
+    );
+    require(
+      account.signerAuthorizations[role][msg.sender].completed == false,
+      "Signer already authorized"
+    );
+
+    authorizedBy[msg.sender] = _account;
+    account.signerAuthorizations[role][msg.sender].completed = true;
+    emit SignerAuthorized(_account, msg.sender, role);
+  }
+
+  /**
+   * @notice Whether or not the signer has been registered as a signer for role
+   * @param account The address of account that authorized signing.
+   * @param signer The address of the signer.
+   * @param role The role that has been authorized.
+   */
+  function isGenericSigner(address account, address signer, string memory role)
+    internal
+    view
+    returns (bool)
+  {
+    require(isAccount(account), "Unknown account");
```

Comment:
I'm not sure we should have a `require` on a view 
This is also redundant in `isDefaultSigner`

Reply:
Why shouldn't we have requires on a view? If it is to simplify certain blocks

Reply:
I suppose it is a question of what is the expected failure mode
Do developers expect to encounter a revert on an `eth_call` or a "default value"?
My point was also that we are performing the same `require` twice in one `isSigner` request

Reply:
Fair enough in this case we can remove the redundant `require`'s. 

In general though we do `require` on view calls a lot, if we now feel it's not the best practice I'm happy to not do it moving forward.

Reply:
We do actually have a test that verifies this call reverts, would we be OK removing that?